### ISSUE TITLE
Assistant Button Detection (LG Devices)

### ISSUE THREAD

**Issue by ksaerdna** [keymapperorg/KeyMapper#497]
Currently the respective hardware key of LG devices is not handled by KeyMapper. However, there is a paid app that can, so it is technically possible. Possibly the solution is the same as with Samsung's pendant.

**Comment by sds100** [comment]
Hello, do you have any idea how the other app does it? I don't have any LG devices or any device with an Assistant button so I have no idea how to approach the problem.

**Comment by ksaerdna** [comment]
If I had had any idea how to technically solve it, I would have told you. :-) I also checked if the source code of the mentioned application is available, but unfortunately it is not. In its basic, free functionality the LG or Samsung button cannot be mapped, instead one needs some kind of in-app purchase.

BTW: All applications I wrote so far for Android are open source, so my intention for this "issue" was not primarily to save a litte amount of money...

**Comment by sds100** [comment]
Yeah. Which app is this? With the Bixby button remapping apps the phone throws an error in the logcat whenever the Bixby button is pressed and the Bixby package is disabled. The user must grant the app permission to read logs through ADB for the feature to work. Can you see if the LG button shows an error in the logcat?

**Comment by ksaerdna** [comment]
I hesitated to place any advertisement, but the application is called _Button Mapper_. Basically it seems to do the same as yours, with the same Android permissions.

LG has three settings regarding the "Assistant Button":

- The "Assistant-App" can be switched between "Google" and "None". No idea if there is a way to configure further ones.
- The Assistant Button can be activated or deactivated
- The Assistant Button longpress can be activated or deactivated

The logcat shows with deactivated Assistant Button:

 HotKeyController: launch Assist: 1 turned off

and with a longpress on lockscreen:

D HotKeyController: AI long action not support : Lockscreen and non-interactive status

and with longpress not on lockscreen, Google configured as Assistant-App:

D HotKeyController: launch Assist: 3 turned off
D HotKeyController: currentAssistComponent : com.google.android.voiceinteraction.GsaVoiceInteractionService
D HotKeyController: received HOTKEY_RELEASE_AFTER_LONG_PRESS: duration=2897
D HotKeyController: launch Assist: 4 turned off

and longpress without "Assistant-App" configured

D HotKeyController: launch Assist: 3 turned off
D HotKeyController: currentAssistComponent :

**Comment by sds100** [comment]
Is there a setup process in the app specific to remapping the Assistant button or does it detect it "out of the box"?

**Comment by sds100** [comment]
Are there log messages for short presses?

**Comment by ksaerdna** [comment]
> Is there a setup process in the app specific to remapping the Assistant button or does it detect it "out of the box"?

I cannot know because I did not follow the in-app purchase request, but as far as I read from users' reports it works "out of the box".

>Are there log messages for short presses?

The first log message I mentioned (**Assist 1**) resulted from a short press. The message is the same, during or without lockscreen. Note that there is ridiculous flood of stupid messages, and finally I used "grep HotKeyController" for filtering.

**Comment by sds100** [comment]
If the app detects the button press by filtering the logcat then it can't work out of the box because an app needs a permission that can only be granted through the shell. Can you find a youtube video or something that can show how it is set up? It isn't a huge deal because the logging information that you gave me is enough to detect when it is pressed.

Was the button "deactivated" for all of those logcat messages? When the button is "deactivated" does it not do anything on the device when you press it? This is important because Key Mapper will not be able to **remap** the button but can detect when it is supposedly pressed and then do something, assuming that nothing else is happening on the device to interfere with the action.

**Comment by sds100** [comment]
This button won't have the same amount of flexibility as normal triggers because I can't detect when it is pressed _down_ and then _up_. Do those logcat messages show up _after_ you release the button? These logcat messages seem more trivial to interpret than the Bixby ones.

**Comment by ksaerdna** [comment]
The command for the mentioned application seems to be:

**adb shell pm grant flar2.homebutton android.permission.READ_LOGS** 

as I saw in the respective xda thread (https://forum.xda-developers.com/android/apps-games/app-button-mapper-customize-hardware-t3361340). There are more "pm grant" commands mentioned, like WRITE_SECURE_SETTINGS and others.

On LG: The HotKeyController messages appear in this context:

**D InputReader: Key event: device=6 type=0x0001 value=0x00000000 when=121391700184000
D WindowManager: interceptKeyTq keycode=165 down=false injected=false mDisplayState=2 interactive=true keyguardActive=false flags=0x8 policyFlags=22000000
D TeleService: OtaspCallManager: isOtaSpActive is false only
D PocketDetection: isInthePocketInternal: isInthePocket = false
D LGPhoneWindowManager: hotKeyRelease count= : 0, duration=608
D LGPhoneWindowManager: hotKeyActions : mode=11, interactive=true, type=1
D HotKeyController: launch Assist: 1 turned off
D LGPhoneWindowManager: hotKeyLongRelease : handled=false, count=0**

So there are more processes involved in the key handling than only HotKeyController.

**Comment by lucom** [comment]
I also own an LG device and have the same problem: The Assistant button cannot be remapped with your tool.
Would it be possible to register your Key Mapper app as an alternative "Assistant App" so that one could select it in the Android settings?

Alternatively: How could I help get this issue fixed?

**Comment by sds100** [comment]
Yes, I think I can add that feature. That would the most trivial way to add support for Assistant buttons. It would be a very simple key map because the Assist app is just told to open and isn't given any information about how the button was pressed.

For the hacky work around can you confirm that you get the same logcat output as the other person? You'll need to figure out how Key Mapper can determine that the assistant button was short pressed or long pressed by just parsing the logcat. The same messages need to be outputted  _**consistently**_  every time you short or long press the button.

**Comment by Susko3** [comment]
**Phone:** LG VELVET

**Relevant settings:** [image](https://user-images.githubusercontent.com/16479013/117585759-6365a480-b114-11eb-918a-a77093f2aaf7.png)

**Logcat, with the last two settings disabled:**

Short press:
```
D HotKeyController: launch Assist: 1 turned off
```

Long press:
```
D HotKeyController: launch Assist: 3 turned off
D HotKeyController: currentAssistComponent : com.google.android.voiceinteraction.GsaVoiceInteractionService
D HotKeyController: received HOTKEY_RELEASE_AFTER_LONG_PRESS: duration=1293
D HotKeyController: launch Assist: 4 turned off
```

"Short" long press:
```
D HotKeyController: launch Assist: 3 turned off
D HotKeyController: currentAssistComponent : com.google.android.voiceinteraction.GsaVoiceInteractionService
D HotKeyController: received HOTKEY_RELEASE_AFTER_LONG_PRESS: duration=22
D HotKeyController: need2moreTime =478
D HotKeyController: launch Assist: 4 turned off
```

Long press (when screen is off and on the lock screen), one entry for press and another for release:
```
D HotKeyController: AI long action not support : Lockscreen and non-interactive status
D HotKeyController: AI long action not support : Lockscreen and non-interactive status
```

It's also possible to get this output if you hold the assistant button, and then turn the screen off:
```
D HotKeyController: launch Assist: 3 turned off
D HotKeyController: AI long action not support : Lockscreen and non-interactive status
```

**Comment by sds100** [comment]
![image](https://user-images.githubusercontent.com/16245954/117586465-521a9900-b110-11eb-9bab-5fbadbbfca99.png)
Do these options still show if you select a 3rd party assist app? like Firefox or DuckDuckGo?

**Comment by IvanBayan** [comment]
I've faced with similar issue on android 11 nokia. Key mapper just didn't pick up google assistant button until I switched it off in gesture settings (Settings->System->Gestures  disable google assistant button).

**Comment by GL513** [comment]
https://forum.xda-developers.com/t/how-to-remap-the-lg-v50-assistant-key-no-root.4029035/

https://forum.xda-developers.com/t/new-project-remap-google-assistant-key.3934930/

These are links i had bookmarked from when i had the G7 thinQ, i dont have it anymore so i cant be of much use other than this.

**Comment by raniesantos** [comment]
Issue status?

I'm on a Nokia G10. I want to turn my assistant button into an instant mute button.

Trigger:
unknown keycode 289 (Nokia's assistant button)

Actions:
Mute volume
Pause media
Lock device

This kind of works BUT after the phone gets locked, a split second later the screen wakes up because Google Assistant appears at the bottom of the screen.

Disabling the button will completely prevent the trigger from being received.

I can't change the default assistant to None because I actually use it and find it convenient.

I think the only solution is if this app can be set as the assistant. And then, some kind of condition needs to be met. For example, **if** the keymapper assistant is activated by this keypress **then** execute these actions, **else** launch Google assistant.

So as an assistant app, keymapper would respond differently based on how it was triggered. The hardware button would execute user defined actions while other ways of being launched would just activate the default assistant.

No idea how complicated this is, I've never done Android development.

Side note: There's an app on the Play Store called assist mapper, it can be set as an assistant app. So I at least know that part of my idea is possible, but I have no clue if the detection of how an app was launched is possible.

**Comment by sudoshindo** [comment]
> The new assistant trigger should fix some of these issues

Which version is it supposed to be in so that we can check?

**Comment by sudoshindo** [comment]
@sds100 I just want a bit of clarification. So the new paid feature will set this app as an "Assistant" on the phone.

And from there, we would be able to **conditionally trigger** the actions similar to what someone else described previously?

> I think the only solution is if this app can be set as the assistant. And then, some kind of condition needs to be met. For example, if the keymapper assistant is activated by this keypress then execute these actions, else launch Google assistant.

**Comment by sudoshindo** [comment]
@sds100 I just want a bit of clarification. So the new paid feature will set this app as an "Assistant" on the phone.

So after setting that, will I be able to do the following:

IF the physical button is pressed
THEN trigger actions.

ELSE IF the assistant is triggered by other means or by software buttons (e.g the Nova launcher widget in the below screenshot)
THEN launch the default Google Assistant

![Image](https://github.com/user-attachments/assets/ec65d338-5cf0-4317-90b9-2bd286611cfb)

If that's the case then I would pay for it as soon as you confirm my assumption.

**Comment by jambl3r** [comment]
If you launch specifically Google Assistant, then it will launch, but by using the assistant trigger you are changing your default assistant to Key Mapper, so given any other opportunity, your key map will launch instead.

Feel free to try out the feature, and refund it if it doesn't work. Any problems with that let me know and I can personally refund it.
